### PR TITLE
[WebAuthn] Place PRF behind guard and remove from WKA

### PR DESCRIPTION
**Pull request by pascoej**
#### e8cdcf9f49d46edeac24df2d888493f6903032ea
<pre>
[WebAuthn] Place PRF behind guard and remove from WKA
<a href="https://bugs.webkit.org/show_bug.cgi?id=276770">https://bugs.webkit.org/show_bug.cgi?id=276770</a>
<a href="https://rdar.apple.com/problem/132098736">rdar://problem/132098736</a>

Reviewed by Brent Fulgham.

The platform PRF API needs to behind guards in order to avoid calling that API
in build configurations that don&apos;t support it.

* Source/WTF/wtf/PlatformHave.h:
* Source/WebKit/UIProcess/WebAuthentication/Cocoa/AuthenticationServicesForwardDeclarations.h:
* Source/WebKit/UIProcess/WebAuthentication/Cocoa/AuthenticationServicesSoftLink.h:
* Source/WebKit/UIProcess/WebAuthentication/Cocoa/AuthenticationServicesSoftLink.mm:
* Source/WebKit/UIProcess/WebAuthentication/Cocoa/WebAuthenticatorCoordinatorProxy.mm:
(WebKit::toASAssertionPRFInputValue):
(WebKit::WebAuthenticatorCoordinatorProxy::requestsForRegistration):
(WebKit::WebAuthenticatorCoordinatorProxy::requestsForAssertion):
(WebKit::WebAuthenticatorCoordinatorProxy::performRequest):

Canonical link: <a href="https://commits.webkit.org/281212@main">https://commits.webkit.org/281212@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/5b05b66f059f4d11c1cc9a9b65fc04ea04120999

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/59117 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/38450 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/11615 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/62941 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/9560 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/46102 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/9776 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/47983 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/6838 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/61147 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/35962 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/51163 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/28819 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/32685 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/8427 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/8564 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/52209 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/54640 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/8708 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/64642 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/58360 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/3036 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/8665 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/55309 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/3047 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/51164 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/55425 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/13092 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/2546 "Passed tests") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/80119 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/34276 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/13869 "Passed tests") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/35359 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/36445 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/35104 "Built successfully") | | |

### CODE REVIEW

File: Source/WebKit/UIProcess/WebAuthentication/Cocoa/AuthenticationServicesForwardDeclarations.h [hunk 1]
```diff
@@ -359,6 +359,59 @@ typedef NSString *ASAuthorizationPublicKeyCredentialResidentKeyPreference;
 
 @end
 
+
+typedef NS_ENUM(NSInteger, ASAuthorizationPlatformPublicKeyCredentialRegistrationRequestStyle) {
+    ASAuthorizationPlatformPublicKeyCredentialRegistrationRequestStyleStandard,
+    ASAuthorizationPlatformPublicKeyCredentialRegistrationRequestStyleConditional,
+};
+
+@interface ASAuthorizationPublicKeyCredentialPRFAssertionOutput: NSObject
```
Comment (pxlcoder) [inline Source/WebKit/UIProcess/WebAuthentication/Cocoa/AuthenticationServicesForwardDeclarations.h]:
```suggestion
@interface ASAuthorizationPublicKeyCredentialPRFAssertionOutput : NSObject
```

Many instances throughout this file.

File: Source/WebKit/UIProcess/WebAuthentication/Cocoa/AuthenticationServicesForwardDeclarations.h [hunk 2]
```diff
@@ -359,6 +359,59 @@ typedef NSString *ASAuthorizationPublicKeyCredentialResidentKeyPreference;
...
+@property (nonatomic, nullable, readonly) NSData *saltInput2;
+@end
+
+@interface ASAuthorizationPublicKeyCredentialPRFAssertionInput: NSObject
+- (instancetype)initWithInputValues:(nullable ASAuthorizationPublicKeyCredentialPRFAssertionInputValues *)inputValues perCredentialInputValues:(nullable NSDictionary<NSData *, ASAuthorizationPublicKeyCredentialPRFAssertionInputValues *> *)perCredentialInputValues;
+@property (nonatomic, nullable, readonly) ASAuthorizationPublicKeyCredentialPRFAssertionInputValues *inputValues;
+@property (nonatomic, nullable, readonly) NSDictionary<NSData *, ASAuthorizationPublicKeyCredentialPRFAssertionInputValues *> *perCredentialInputValues;
+@end
+
+@interface ASAuthorizationPlatformPublicKeyCredentialAssertionRequest (Staging)
```
Comment (pxlcoder) [inline Source/WebKit/UIProcess/WebAuthentication/Cocoa/AuthenticationServicesForwardDeclarations.h]: We only use the term "Staging" if the code is intended to be removed at some point. If not, I would just use:

```
@interface ASAuthorizationPlatformPublicKeyCredentialAssertionRequest ()
```

Same comment for the other instances.

File: Source/WebKit/UIProcess/WebAuthentication/Cocoa/WebAuthenticatorCoordinatorProxy.mm [hunk 3]
```diff
@@ -299,7 +307,17 @@ static inline ASAuthorizationPublicKeyCredentialLargeBlobSupportRequirement toAS
     if (includePlatformRequest) {
         RetainPtr request = adoptNS([[allocASAuthorizationPlatformPublicKeyCredentialProviderInstance() initWithRelyingPartyIdentifier:*options.rp.id] createCredentialRegistrationRequestWithClientData:clientData.get() name:options.user.name userID:toNSData(options.user.id).get()]);
 
-        AUTHENTICATOR_COORDINATOR_ADDITIONS
+        if (m_isConditionalMediation && [request.get() respondsToSelector:@selector(setRequestStyle:)])
```
Comment (pxlcoder) [inline Source/WebKit/UIProcess/WebAuthentication/Cocoa/WebAuthenticatorCoordinatorProxy.mm]:
```suggestion
        if (m_isConditionalMediation && [request respondsToSelector:@selector(setRequestStyle:)])
```

File: Source/WebKit/UIProcess/WebAuthentication/Cocoa/WebAuthenticatorCoordinatorProxy.mm [hunk 4]
```diff
@@ -405,7 +423,23 @@ static inline bool allowsHybrid(const Vector<AuthenticatorTransport>& transports
         if (!allowHybrid)
             request.get().shouldShowHybridTransport = false;
 
-        AUTHENTICATOR_COORDINATOR_ASSERTION_REQUEST_ADDITIONS
+#if HAVE(WEB_AUTHN_PRF_API)
+        if (options.extensions && options.extensions->prf) {
+            auto prf = options.extensions->prf;
+            RetainPtr<ASAuthorizationPublicKeyCredentialPRFAssertionInputValues> inputValues = toASAssertionPRFInputValue(prf->eval);
```
Comment (pxlcoder) [inline Source/WebKit/UIProcess/WebAuthentication/Cocoa/WebAuthenticatorCoordinatorProxy.mm]:
```suggestion
            RetainPtr inputValues = toASAssertionPRFInputValue(prf->eval);
```

File: Source/WebKit/UIProcess/WebAuthentication/Cocoa/WebAuthenticatorCoordinatorProxy.mm [hunk 5]
```diff
@@ -405,7 +423,23 @@ static inline bool allowsHybrid(const Vector<AuthenticatorTransport>& transports
...
+            RetainPtr<NSMutableDictionary> perCredentialInputValues = nullptr;
+            if (prf->evalByCredential) {
+                perCredentialInputValues = adoptNS([[NSMutableDictionary alloc] init]);
+                for (auto& credentialIDAndInputValues : *prf->evalByCredential) {
+                    auto key = base64URLDecode(credentialIDAndInputValues.key.utf8().span());
+                    if (!key)
+                        continue;
+                    [perCredentialInputValues.get() setObject:toASAssertionPRFInputValue(credentialIDAndInputValues.value).get() forKey: toNSData(*key).get()];
```
Comment (pxlcoder) [inline Source/WebKit/UIProcess/WebAuthentication/Cocoa/WebAuthenticatorCoordinatorProxy.mm]:
```suggestion
                    [perCredentialInputValues setObject:toASAssertionPRFInputValue(credentialIDAndInputValues.value).get() forKey: toNSData(*key).get()];
```

File: Source/WebKit/UIProcess/WebAuthentication/Cocoa/WebAuthenticatorCoordinatorProxy.mm [hunk 6]
```diff
@@ -558,7 +592,21 @@ static inline bool allowsHybrid(const Vector<AuthenticatorTransport>& transports
                     extensionOutputs.largeBlob = { credential.get().largeBlob.isSupported, nullptr, std::nullopt };
                 }
 
-                AUTHENTICATOR_COORDINATOR_REGISTRATION_RESPONSE_ADDITIONS
+#if HAVE(WEB_AUTHN_PRF_API)
+                if ([credential.get() respondsToSelector:@selector(prf)] && credential.get().prf) {
```
Comment (pxlcoder) [inline Source/WebKit/UIProcess/WebAuthentication/Cocoa/WebAuthenticatorCoordinatorProxy.mm]:
```suggestion
                if ([credential respondsToSelector:@selector(prf)] && credential.get().prf) {
```

File: Source/WebKit/UIProcess/WebAuthentication/Cocoa/WebAuthenticatorCoordinatorProxy.mm [hunk 7]
```diff
@@ -580,7 +628,16 @@ static inline bool allowsHybrid(const Vector<AuthenticatorTransport>& transports
                     extensionOutputs.largeBlob = { std::nullopt, protector, credential.get().largeBlob.didWrite };
                 }
 
-                AUTHENTICATOR_COORDINATOR_ASSERTION_RESPONSE_ADDITIONS
+#if HAVE(WEB_AUTHN_PRF_API)
+                if ([credential.get() respondsToSelector:@selector(prf)] && credential.get().prf) {
```
Comment (pxlcoder) [inline Source/WebKit/UIProcess/WebAuthentication/Cocoa/WebAuthenticatorCoordinatorProxy.mm]:
```suggestion
                if ([credential respondsToSelector:@selector(prf)] && credential.get().prf) {
```

File: Source/WebKit/UIProcess/WebAuthentication/Cocoa/AuthenticationServicesForwardDeclarations.h [hunk 8]
```diff
@@ -359,6 +359,58 @@ typedef NSString *ASAuthorizationPublicKeyCredentialResidentKeyPreference;
 
 @end
 
+typedef NS_ENUM(NSInteger, ASAuthorizationPlatformPublicKeyCredentialRegistrationRequestStyle) {
```
Comment (brentfulgham) [inline Source/WebKit/UIProcess/WebAuthentication/Cocoa/AuthenticationServicesForwardDeclarations.h:362]: Shouldn't this section be protected by `HAVE(WEB_AUTHN_PRF_API)`?

File: Source/WebKit/UIProcess/WebAuthentication/Cocoa/AuthenticationServicesSoftLink.h [hunk 9]
```diff
@@ -47,4 +47,7 @@ SOFT_LINK_CLASS_FOR_HEADER(WebKit, ASAuthorizationSecurityKeyPublicKeyCredential
 SOFT_LINK_CLASS_FOR_HEADER(WebKit, ASAuthorizationPublicKeyCredentialLargeBlobAssertionInput);
 SOFT_LINK_CLASS_FOR_HEADER(WebKit, ASAuthorizationPublicKeyCredentialLargeBlobRegistrationInput);
 SOFT_LINK_CONSTANT_FOR_HEADER(WebKit, AuthenticationServices, ASAuthorizationErrorDomain, NSErrorDomain);
+SOFT_LINK_CLASS_FOR_HEADER(WebKit, ASAuthorizationPublicKeyCredentialPRFRegistrationInput);
```
Comment (brentfulgham) [inline Source/WebKit/UIProcess/WebAuthentication/Cocoa/AuthenticationServicesSoftLink.h:51]: Ditto.

File: Source/WebKit/UIProcess/WebAuthentication/Cocoa/AuthenticationServicesSoftLink.mm [hunk 10]
```diff
@@ -43,3 +43,6 @@
 SOFT_LINK_CLASS_FOR_SOURCE(WebKit, AuthenticationServices, ASAuthorizationPublicKeyCredentialLargeBlobAssertionInput);
 SOFT_LINK_CLASS_FOR_SOURCE(WebKit, AuthenticationServices, ASAuthorizationPublicKeyCredentialLargeBlobRegistrationInput);
 SOFT_LINK_CONSTANT_FOR_SOURCE(WebKit, AuthenticationServices, ASAuthorizationErrorDomain, NSErrorDomain);
+SOFT_LINK_CLASS_FOR_SOURCE(WebKit, AuthenticationServices, ASAuthorizationPublicKeyCredentialPRFRegistrationInput);
```
Comment (brentfulgham) [inline Source/WebKit/UIProcess/WebAuthentication/Cocoa/AuthenticationServicesSoftLink.mm:47]: Ditto.

**Review by brentfulgham — APPROVED** [review]
Please guard the forward declarations with the macro, since at least the implementation portion of the SoftLink.mm code generates stuff we have to build and that's just wasted binary size and compile time.

**Comment by brentfulgham** [comment]
Note: The WPE failures cannot be related to this Cocoa-specific file change.

**Review by brentfulgham — APPROVED** [review]

**Pull request merged**